dali/kernels/span.h

```diff
@@ -0,0 +1,182 @@
+// Copyright (c) 2017-2018, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_SPAN_H_
+#define DALI_KERNELS_SPAN_H_
+
+#include <cstddef>
+#include <type_traits>
+
+namespace tensor {
+
+// Based on "span: bounds-safe views for sequences of objects"
+// http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0122r7.pdf
+// Adopted to c++20
+// Missing: containers constructors, static <-> dynamic, reverse iterators, subspans, comparisions
+// Removed some constexprs due to c++11 limitations
+
+constexpr ptrdiff_t dynamic_extent = -1;
+
+template <class ElementType, ptrdiff_t Extent = dynamic_extent>
+class span {
+ public:
+  // constants and types
+  using element_type = ElementType;
+  using value_type = typename std::remove_cv<ElementType>::type;
+  using index_type = ptrdiff_t;
+  using difference_type = ptrdiff_t;
```

Comment:
Why not use directly `ptrdiff_t` here?

Reply:
difference_type can be used by some STL algorithms (at least I think so).